packages/loopring_v3/contracts/amm/libamm/AmmUpdateProcess.sol

```diff
@@ -54,13 +55,19 @@ library AmmUpdateProcess
                 // update.tokenID == token.tokenID &&
                 // update.feeBips == ctx.feeBips &&
                 packedDataA & 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffff ==
-                (uint(ExchangeData.TransactionType.AMM_UPDATE) << 216) | (uint(address(this)) << 56) | (ctx.accountID << 24) | (token.tokenID << 8) | ctx.feeBips &&
-                // update.tokenWeight == token.weight
-                (packedDataB >> 128) & 0xffffffffffffffffffffffff == token.weight,
+                (uint(ExchangeData.TransactionType.AMM_UPDATE) << 216) | (uint(address(this)) << 56) | (ctx.accountID << 24) | (token.tokenID << 8) | ctx.feeBips,
                 "INVALID_AMM_UPDATE_TX_DATA"
             );
 
-            ctx.tokenBalancesL2[i] = /*tokenWeight*/uint96(packedDataB & 0xffffffffffffffffffffffff);
+            if (start) {
+                ctx.tokenBalancesL2[i] = /*balance*/uint96(packedDataB & 0xffffffffffffffffffffffff);
+                ctx.vTokenBalancesL2[i] = /*vbalance*/uint96((packedDataB >> 128) & 0xffffffffffffffffffffffff);
+            } else {
+                require(
+                    ctx.vTokenBalancesL2[i] == /*vbalance*/uint96((packedDataB >> 128) & 0xffffffffffffffffffffffff),
```

Comment:
We update these on L1 in the exact same way the balances are updated on L2 using transfers (so on L1 we use the amount transferred as provided in the DA), so they have to be the same, otherwise there is a bug somewhere.

This is unlike the virtual balances, which can be updated to any value on L1.